kong/plugins/ldap-auth/asn1.lua

```diff
@@ -1,408 +1,305 @@
+    return ffi_string(charpp[0], ret)
+  end
 
-      local encVal = ""
-      for _, v in ipairs(val) do
-        encVal = encVal .. self.encode(v) -- todo: buffer?
-      end
+  -- Octet String
+  encoder[TAG.OCTET_STRING] = function(val)
+    local typ = C.ASN1_OCTET_STRING_new()
+    C.ASN1_STRING_set(typ, val, #val)
+    charpp[0] = nil
+    local ret = C.i2d_ASN1_OCTET_STRING(typ, charpp)
+    C.ASN1_STRING_free(typ)
+    return ffi_string(charpp[0], ret)
+  end
 
-      local tableType = "\x30"
-      if val["_snmp"] then
-        tableType = bpack("X", val["_snmp"])
-      end
+  encoder[TAG.ENUMERATED] = function(val)
+    local typ = C.ASN1_ENUMERATED_new()
+    C.ASN1_ENUMERATED_set(typ, val)
+    charpp[0] = nil
+    local ret = C.i2d_ASN1_ENUMERATED(typ, charpp)
+    C.ASN1_INTEGER_free(typ)
+    return ffi_string(charpp[0], ret)
+  end
 
-      return bpack("AAA", tableType, self.encodeLength(#encVal), encVal)
-    end
+  encoder[TAG.SEQUENCE] = function(val)
+    return asn1_put_object(TAG.SEQUENCE, CLASS.UNIVERSAL, 1, val)
+  end
 
-    -- Boolean encoder
-    self.encoder["boolean"] = function(self, val)
-      if val then
-        return bpack("X", "01 01 FF")
-      else
-        return bpack("X", "01 01 00")
+  function encode(val, tag)
+    if tag == nil then
+      local typ = type(val)
+      if typ == "string" then
+        tag = TAG.OCTET_STRING
+      elseif typ == "number" then
+        tag = TAG.INTEGER
       end
     end
 
-    -- Integer encoder
-    self.encoder["number"] = function(self, val)
-      local ival = self.encodeInt(val)
-      local len = self.encodeLength(#ival)
-
-      return bpack("XAA", "02", len, ival)
+    if encoder[tag] then
+      return encoder[tag](val)
     end
+  end
+end
+_M.encode = encode
 
-    -- Octet String encoder
-    self.encoder["string"] = function(self, val)
-      local len = self.encodeLength(#val)
-      return bpack("XAA", "04", len, val)
-    end
-
-    -- Null encoder
-    self.encoder["nil"] = function(self, val)
-      return bpack("X", "05 00")
-    end
-  end,
-
-  encode_oid_component = function(n)
-    local parts = {}
-
-    parts[1] = char(n % 128)
-    while n >= 128 do
-      n = bit.rshift(n, 7)
-      parts[#parts + 1] = char(n % 128 + 0x80)
-    end
-
-    return reverse(concat(parts))
-  end,
-
-  encodeInt = function(val)
-    local lsb = 0
-
-    if val > 0 then
-      local valStr = ""
-
-      while (val > 0) do
-        lsb = math.fmod(val, 256)
-        valStr = valStr .. bpack("C", lsb)
-        val = math.floor(val/256)
-      end
-
-      if lsb > 127 then
-        valStr = valStr .. "\0"
-      end
-
-      return reverse(valStr)
-
-    elseif val < 0 then
-      local i = 1
-      local tcval = val + 256
 
-      while tcval <= 127 do
-        tcval = tcval + (math.pow(256, i) * 255)
-        i = i+1
-      end
+local decode
+do
+  local decoder = new_tab(0, 2)
 
-      local valStr = ""
+  decoder[TAG.OCTET_STRING] = function(der, offset, len)
```

Comment:
let's add an assert here, even if the caller may have checked the size
`assert(offset < #der)`

Reply:
Thanks!